include/revng/StackAnalysis/StackAnalysis.h

```diff
@@ -52,4 +76,201 @@ class StackAnalysis : public llvm::ModulePass {
+      return std::includes(Old.ClobberedRegisters.begin(),
+                           Old.ClobberedRegisters.end(),
+                           New.ClobberedRegisters.begin(),
+                           New.ClobberedRegisters.end());
+
+    return New.Type <= Old.Type;
+  }
+
+  void dump() const debug_function { dump(dbg); }
+
+  template<typename T>
+  void dump(T &Output) const {
+    Output << "Dumping summary \n"
+           << "  Type: " << Type << "\n"
+           << "  ElectedFSO: " << (ElectedFSO.has_value() ? *ElectedFSO : -1)
+           << "\n"
+           << "  Clobbered registers: \n";
+
+    for (auto *Reg : ClobberedRegisters)
+      Output << "    " << Reg->getName().str() << "\n";
+
+    Output << "  ABI info: \n";
+    ABIResults.dump();
+  }
+};
+
+class FunctionAnalysisResults {
+private:
+  /// \brief Map from CFEP to its function description
+  std::map<MetaAddress, FunctionSummary> FunctionsBucket;
+  FunctionSummary DefaultSummary;
+
+public:
+  FunctionAnalysisResults(FunctionSummary DefaultSummary) :
+    DefaultSummary(std::move(DefaultSummary)) {}
+
+  FunctionSummary &at(MetaAddress PC) { return FunctionsBucket.at(PC); }
+
+  const FunctionSummary &at(MetaAddress PC) const {
+    return FunctionsBucket.at(PC);
+  }
+
+  model::FunctionType::Values getFunctionType(MetaAddress PC) const {
+    return get(PC).Type;
+  }
+
+  bool isFakeFunction(MetaAddress PC) const {
+    return getFunctionType(PC) == model::FunctionType::Values::Fake;
+  }
+
+  llvm::Function *getFakeFunction(MetaAddress PC) const {
+    return get(PC).FakeFunction;
+  }
+
+  const auto &getRegistersClobbered(MetaAddress PC) const {
+    return get(PC).ClobberedRegisters;
+  }
+
+  std::optional<uint64_t> getElectedFSO(MetaAddress PC) const {
+    return get(PC).ElectedFSO;
+  }
+
+  bool registerFunction(MetaAddress PC, FunctionSummary &&F) {
+    revng_assert(PC.isValid());
+    auto It = FunctionsBucket.find(PC);
+    if (It != FunctionsBucket.end()) {
+      bool Changed = FunctionSummary::compare(It->second, F);
+      It->second = std::move(F);
+      return not Changed;
+    } else {
+      FunctionsBucket.emplace(PC, std::move(F));
+      return true;
+    }
+  }
+
+private:
+  const FunctionSummary &get(MetaAddress PC) const {
+    auto It = FunctionsBucket.find(PC);
+    if (It != FunctionsBucket.end())
+      return It->second;
+    return DefaultSummary;
+  }
+};
+
+template<class FunctionOracle>
+class CFEPAnalyzer {
+private:
+  llvm::Module &M;
+  llvm::LLVMContext &Context;
+  GeneratedCodeBasicInfo *GCBI;
+  FunctionOracle &Oracle;
+  llvm::Function *PreHookMarker;
+  llvm::Function *PostHookMarker;
+  llvm::Function *RetHookMarker;
+  llvm::Function *IndirectBranchInfoMarker;
+  llvm::Function *UnexpectedPCMarker;
+  llvm::raw_fd_ostream &OutputIBI;
+  llvm::raw_fd_ostream &OutputAAWriter;
+  OpaqueFunctionsPool<llvm::StringRef> RegistersClobberedPool;
+  llvm::BasicBlock *AnyPCCloned;
+  llvm::BasicBlock *UnexpectedPCCloned;
```

Comment:
**ALL** the scalar fields (i.e., integers, bool and **pointers**) must always be zero-initialized or initialized in *all* constructors.